src/main/java/com/android/volley/toolbox/CacheHeader.java

```diff
@@ -115,6 +120,40 @@ static CacheHeader readHeader(DiskBasedCache.CountingInputStream is) throws IOEx
                 key, etag, serverDate, lastModified, ttl, softTtl, allResponseHeaders);
     }
 
+    /**
+     * Reads the header from a AsynchronousFileChannel and returns a CacheHeader object.
+     *
+     * @param afc File channel to read from.
+     * @param size Size of the cache header.
+     * @throws IOException if fails to read header
+     */
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    @Nullable
+    static CacheHeader readHeader(final AsynchronousFileChannel afc, int size) throws IOException {
```

Comment:
Several problems here:
1. Declares that it throws IOException but returns null instead? Static analysis (e.g. FindBugs) should flag this?
2. All exceptions are hidden without a log message.
3. Why throw IOE if you're going to return null anyway?
4. What about interrupted exception?

The smallest fix is to move the try-catch to contain only the future.get().
Handle interrupted and execution exception there:
* The cause of the execution exception is very likely to be IOException. I suggest rethrowing the cause wrapped in a new IOException at this point.
* How to handle interrupted exception? Rethrow as IOException is one idea.

Then remove the outer try-catch and let IOException in body throw from method normally.

CompletionHandler here would allow you to compose this operation and postpone these decisions?

Reply:
I'm a little confused, since I thought the original issue was rethrowing the exceptions as IOExceptions? I just updated the code, so that the try-catch only contains future.get(), logs both exceptions so they aren't hidden, and then rethrows as an IOException. Does this fix your concerns?

I had tried to use a completion handler, but the code seemed to be a good bit more complicated.

Reply:
Sorry for the run-around. On closer inspection I see that my InterruptedException "suggestion" was incorrect. This InterruptedException happens if the caller to `future.get()` is interrupted while waiting. This is unlikely if the caller is running on the main thread.

One typical way to handle this is to reassert the interrupt so it's not swallowed. I think that's what you should do. (Check out the chapter in Effective Java for more guidance.)

Reply:
okay, thanks!

Reply:
Sorry for joining this discussion late!

I think the reason we're running into pain here is because we're doing blocking I/O in a non-blocking API implementation. The aim of the asynchronous API is that we should always return immediately, without blocking on any of the disk operations, including reading of the cache header. Therefore, we shouldn't be making a blocking call to get() inside the asynchronous get method. If we're doing any sort of operation that can throw an InterruptedException in an asynchronous method, that's usually not a good sign.

We should start by reading the header asynchronously, and chain a callback once that read completes to either handle invalid headers or else perform the rest of the read, and then dispatch the user's callback once it completes. This can be done with nested callbacks, or by chaining futures if supported at this API level.